Plugin.pm

```diff
@@ -628,6 +635,164 @@ sub getFeatured {
+
+		my $items = [ map {
+			{
+				name => $_->{title},
+				type => 'link',
+				url => $_->{type} eq 'HIGHLIGHT_MODULE' ? \&getHighlights : \&getModule,
+				passthrough => [ { module => $_ } ],
+			}
+		} @$modules ];
+
+		$cb->( {
+			items => $items
+		} );
+	}, 'pages/home' );
+}
+
+sub getHighlights {
+	my ( $client, $cb, $args, $params ) = @_;
+
+	my $module = $params->{module};
+	my $items = [];
+
+	foreach my $entry (@{$module->{highlights}}) {
+		my $item;
+		my $title = $entry->{title};
+
+		$entry = $entry->{item};
+
+		# formats are not exactly consistent with the rest of the API. Small changes
+		# but requires a few exceptions. We could change getImageUrl and typeOfItem...
+		if ($entry->{type} eq 'ALBUM') {
+			$item = _renderAlbum($entry->{item}, { addArtistToTitle => 1 });
+		}
+		elsif ($entry->{type} eq 'PLAYLIST') {
+			$item = _renderPlaylist($entry->{item});
+			# image can't be 1280x1280...
+			my $image = $entry->{item}->{squareImage} =~ s/-/\//gr;
+			$item->{image} = IURL . $image . "/640x640.jpg";
+		}
+		elsif ($entry->{type} =~ /MIX/) {
+			$item = _renderMix($client, $entry->{item});
+			# image uses a different key (S/M/L) than current MIX
+			$item->{image} = $entry->{item}->{images}->{MEDIUM}->{url} || $entry->{item}->{images}->{SMALL}->{url} || $entry->{item}->{images}->{LARGE}->{url};;
+		}
+		elsif ($entry->{type} eq 'TRACK') {
+			# I don't think we should cache it in Async b/c that means a lot of knowledge of the format
+			my ($track) = @{Plugins::TIDAL::API->cacheTrackMetadata([ $entry->{item} ])};
+			$item = _renderTrack($track, { addArtistToTitle => 1 });
+		}
+		else {
+			$log->warn("unknow hightlight type: $entry->{type}");
+		}
+
+		$item->{name} = "$title: $item->{name}" unless $entry->{type} eq 'MIX';
+		push @$items, $item;
+	}
+
+	$cb->({
+		image => 'plugins/TIDAL/html/personal.png',
+		items => $items,
+	});
+}
+
+sub getModule {
+	my ( $client, $cb, $args, $params ) = @_;
+
+	my $module = $params->{module};
+	my $items = _renderModule($client, $module->{type}, $module->{pagedList}->{items});
+
+	# don't ask for more if we have all items
+	unshift @$items, {
+		name => $module->{showMore}->{title},
+		type => 'link',
+		image => __PACKAGE__->_pluginDataFor('icon'),
+		url => \&getDataPage,
+		passthrough => [ {
+			page => $module->{pagedList}->{dataApiPath},
+			limit => $module->{pagedList}->{totalNumberOfItems},
+			type => $module->{type},
+		} ],
+	} if $module->{showMore} && @$items < $module->{pagedList}->{totalNumberOfItems};
+
+	$cb->({
+		items => $items
+	});
+}
+
+sub getDataPage {
+	my ( $client, $cb, $args, $params ) = @_;
+
+	getAPIHandler($client)->dataPage(sub {
+		my $result = shift;
+
+		my $items = _renderModule($client, $params->{type}, $result);
+
+		$cb->({
+			items => $items
+		});
+	}, $params->{page}, $params->{limit} );
+}
+
+sub _renderModule {
+	my ( $client, $type, $entries ) = @_;
+
+	my $items = [];
+
+	if ($type eq 'MIX_LIST') {
+		$items = [ map {
+			{
```

Comment:
Isn't this the same as `_renderMix()` etc.?

Reply:
Yes, they are very close. But I'd need to modify the image and also the typeOfItem detection and I'm afraid to screw the whole rest... Although some keys are missing and I'm not what will be the impact, but that I can try. The issue is really the 2 above